record_indexer/create_collection_index.py

```diff
@@ -0,0 +1,123 @@
+import argparse
+from datetime import datetime
+import json
+import os
+import sys
+
+import boto3
+import requests
+
+from .add_page_to_index import add_page
+from . import settings
+
+
+def update_alias_for_collection(alias: str, collection_id: str, index: str):
+    remove_collection_indices_from_alias(alias, collection_id)
+
+    url = f"{settings.ENDPOINT}/_aliases"
+    headers = {"Content-Type": "application/json"}
+
+    data = {"actions": [{"add": {"index": index, "alias": alias}}]}
+
+    r = requests.post(url, headers=headers, data=json.dumps(data), auth=settings.AUTH)
+    r.raise_for_status()
+    print(f"added index `{index}` to alias `{alias}`")
+
+
+def remove_collection_indices_from_alias(alias: str, collection_id: str):
+    url = f"{settings.ENDPOINT}/_alias/{alias}"
+    r = requests.get(url=url, auth=settings.AUTH)
+    if r.status_code == 404:
+        return
+    else:
+        r.raise_for_status()
+        indices = json.loads(r.text)
+        indices_to_remove = [
+            key for key in indices if key.startswith(f"rikolti-{collection_id}-")
+        ]
+
+        if len(indices_to_remove) > 0:
+            url = f"{settings.ENDPOINT}/_aliases"
+            headers = {"Content-Type": "application/json"}
+            data = {
+                "actions": [{"remove": {"indices": indices_to_remove, "alias": alias}}]
+            }
+
+            r = requests.post(
+                url, headers=headers, data=json.dumps(data), auth=settings.AUTH
+            )
+            r.raise_for_status()
+            print(f"removed indices `{indices_to_remove}` from alias `{alias}`")
+
+
+def delete_old_collection_indices(collection_id: str, retain: int = 1):
+    """
+    Deletes older unaliased indices, retaining a specified number
+    """
+    url = f"{settings.ENDPOINT}/rikolti-{collection_id}-*"
+    params = {"ignore_unavailable": "true"}
+    r = requests.get(url=url, params=params, auth=settings.AUTH)
+    r.raise_for_status()
+    indices = json.loads(r.text)
+
+    unaliased_indices = {}
+    for index in indices.keys():
+        if not indices[index]["aliases"]:
+            creation_date = indices[index]["settings"]["index"]["creation_date"]
+            unaliased_indices[creation_date] = index
+
+    counter = 0
+    for date in reversed(sorted(unaliased_indices)):
+        counter += 1
+        if counter > retain:
+            delete_index(unaliased_indices[date])
```

Comment:
I thought about this for a while, and I've come to the conclusion that in order to guard against a failed index being retained in preference to a good index, we'd have to either:

1. Have a `cleanup_failed_index_task` task that runs after the `create_stage_index_task` task. This would delete a failed index and thus guard against it being retained in future.
2. Upon successful creation of an index, flag it as "good" somewhere (in the Registry DB? Or in another OpenSearch index?). The index deletion algorithm would check this flag, and delete any non-good indices ahead of good ones.

I think option 1 should be fine. I can't think of a reason to keep failed/partial collection indices around. What do you think?

Reply:
And yeah, agreed about making the number of retained indices a setting. I added the INDEX_RETENTION env var.

Reply:
@barbarahui Option 1 sounds like a great use case for the [`@task.branch` operator](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/dags.html#branching). Essentially, if the indexing part succeeds, it returns True, if it fails it returns False, and the next task is a branching task that evaluates if it should clean up failed index, or older unaliased indices.